pkg/openthread/contrib/platform_radio.c

```diff
@@ -0,0 +1,405 @@
+#include "ot.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#include "errno.h"
+#include "net/ethernet/hdr.h"
+#include "net/ethertype.h"
+#include "byteorder.h"
+#include <string.h>
+
+
+
+static RadioPacket sTransmitFrame;
+static RadioPacket sReceiveFrame;
+
+static netdev2_t *_dev;
+
+/* asks the driver the current 15.4 channel */
+uint16_t get_channel(void)
+{
+    uint16_t channel;
+
+    _dev->driver->get(_dev, NETOPT_CHANNEL, &channel, sizeof(uint16_t));
+    return channel;
+}
+
+/* set 15.4 channel */
+int set_channel(uint16_t channel)
+{
+    return _dev->driver->set(_dev, NETOPT_CHANNEL, &channel, sizeof(uint16_t));
+}
+
+/*get transmission power from driver */
+int16_t get_power(void)
+{
+    int16_t power;
+
+    _dev->driver->get(_dev, NETOPT_TX_POWER, &power, sizeof(int16_t));
+    return power;
+}
+
+/* set transmission power */
+int set_power(int16_t power)
+{
+    return _dev->driver->set(_dev, NETOPT_TX_POWER, &power, sizeof(int16_t));
+}
+
+/* set IEEE802.15.4 PAN ID */
+int set_panid(uint16_t panid)
+{
+    return _dev->driver->set(_dev, NETOPT_NID, &panid, sizeof(uint16_t));
+}
+
+/* set extended HW address */
+int set_long_addr(uint8_t *ext_addr)
+{
+    return _dev->driver->set(_dev, NETOPT_ADDRESS_LONG, ext_addr, IEEE802154_LONG_ADDRESS_LEN);
+}
+
+/* set short address */
+int set_addr(uint16_t addr)
+{
+    return _dev->driver->set(_dev, NETOPT_ADDRESS, &addr, sizeof(uint16_t));
+}
+
+/* check the state of promiscuous mode */
+netopt_enable_t is_promiscuous(void)
+{
+    netopt_enable_t en;
+
+    _dev->driver->get(_dev, NETOPT_PROMISCUOUSMODE, &en, sizeof(en));
+    return en == NETOPT_ENABLE ? true : false;;
+}
+
+/* set the state of promiscuous mode */
+int set_promiscuous(netopt_enable_t enable)
+{
+    return _dev->driver->set(_dev, NETOPT_PROMISCUOUSMODE, &enable, sizeof(enable));
+}
+
+/* wrapper for getting device state */
+int get_state(void)
+{
+    netopt_state_t en;
+
+    _dev->driver->get(_dev, NETOPT_STATE, &en, sizeof(netopt_state_t));
+    return en;
+}
+
+/* wrapper for setting device state */
+void set_state(netopt_state_t state)
+{
+    _dev->driver->set(_dev, NETOPT_STATE, &state, sizeof(netopt_state_t));
+}
+
+/* checks if the device is off */
+bool is_off(void)
+{
+    return get_state() == NETOPT_STATE_OFF;
+}
+
+/* sets device state to OFF */
+void ot_off(void)
+{
+    set_state(NETOPT_STATE_OFF);
+}
+
+/* sets device state to SLEEP */
+void ot_sleep(void)
+{
+    set_state(NETOPT_STATE_SLEEP);
+}
+
+/* check if device state is IDLE */
+bool is_idle(void)
+{
+    return get_state() == NETOPT_STATE_IDLE;
+}
+
+/* set device state to IDLE */
+void ot_idle(void)
+{
+    set_state(NETOPT_STATE_IDLE);
+}
+
+/* check if device is receiving a packet */
+bool is_rx(void)
+{
+    return get_state() == NETOPT_STATE_RX;
+}
+
+/* turn on packet reception */
+void enable_rx(void)
+{
+    netopt_enable_t enable = true;
+
+    _dev->driver->set(_dev, NETOPT_RX_LISTENING, &enable, sizeof(enable));
```

Comment:
Hi Jonas.

In the current pkg version of OpenThread there's a race condition between Idle state and packet arrival.
When the driver is in NETOPT_STATE_IDLE, is listening to packets.

When OpenThread sends packets, it calls otPlatRadioIdle first and then otPlatRadioTransmit. From the first call it expects the driver to be Idle and with receiver off. If there's a packet arrival between otPlatRadioIdle and otPlatRadioTransmit, otPlatRadioTransmit will through kThreadError_Busy. This case shouldn't be possible in the OT implementation, so it executes an [assert(false)] and crashes. Althoughthis could be fixed by simulating driver states (and dropping packets), I added this option for easily enabling/disabling the receiver and making it work properly.

The new implementation of OpenThread doesn't have this race condition (otPlatRadioIdle is not needed anymore), so it shouldn't be necessary to implement this RX_LISTENING option (I don't know if it's useful for other applications). I'm currently working in the pkg update to deal with this.

Best.

[assert(false)]: https://github.com/openthread/openthread/blob/master/src/core/mac/mac.cpp#L489